Duplicati/Library/Backend/TencentCOS/COSBackend.cs

```diff
@@ -0,0 +1,421 @@
+                Get(remotename, fs);
+        }
+
+        public void Delete(string remotename)
+        {
+            try
+            {
+                _cosXml = GetCosXml();
+                string bucket = _cosOptions.Bucket; //存储桶，格式：BucketName-APPID
+                string key = GetFullKey(remotename); //对象在存储桶中的位置，即称对象键
+
+                DeleteObjectRequest request = new DeleteObjectRequest(bucket, key);
+
+                //设置签名有效时长
+                request.SetSign(TimeUtils.GetCurrentTime(TimeUnit.SECONDS), 600);
+
+                //执行请求
+                DeleteObjectResult result = _cosXml.DeleteObject(request);
+
+                //请求成功
+                //result.GetResultInfo()
+            }
+            catch (COSXML.CosException.CosClientException clientEx)
+            {
+                Logging.Log.WriteErrorMessage(LOGTAG, "Delete", clientEx, "Delete failed: {0}", remotename);
+                throw;
+            }
+            catch (COSXML.CosException.CosServerException serverEx)
+            {
+                Logging.Log.WriteErrorMessage(LOGTAG, "Delete", serverEx, "Delete failed: {0}, {1}", remotename, serverEx.GetInfo());
+                throw;
+            }
+        }
+
+        public void Test()
+        {
+            var json = JsonConvert.SerializeObject(_cosOptions);
+            try
+            {
+                _cosXml = GetCosXml();
+                string bucket = _cosOptions.Bucket;
+                HeadBucketRequest request = new HeadBucketRequest(bucket);
+                request.SetSign(TimeUtils.GetCurrentTime(TimeUnit.SECONDS), 600);
+                HeadBucketResult result = _cosXml.HeadBucket(request);
+
+                Logging.Log.WriteInformationMessage(LOGTAG, "Test", "Request complete {0}: {1}, {2}", result.httpCode, json, result.GetResultInfo());
+            }
+            catch (COSXML.CosException.CosClientException clientEx)
+            {
+                Logging.Log.WriteErrorMessage(LOGTAG, "Test", clientEx, "Request failed: {0}", json);
+                throw;
+            }
+            catch (COSXML.CosException.CosServerException serverEx)
+            {
+                Logging.Log.WriteErrorMessage(LOGTAG, "Test", serverEx, "Request failed: {0}, {1}", json, serverEx.GetInfo());
+                throw;
+            }
+        }
+
+        public void CreateFolder()
+        {
+            // No need to create folders
+        }
+
+        public void Dispose()
+        {
+            if (_cosXml != null)
+                _cosXml = null;
+        }
+
+        public Task PutAsync(string remotename, Stream stream, CancellationToken cancelToken)
+        {
+            try
+            {
+                _cosXml = GetCosXml();
+                string bucket = _cosOptions.Bucket;
+                string key = GetFullKey(remotename);
+
+                byte[] bytes = new byte[stream.Length];
+                stream.Read(bytes, 0, bytes.Length);
```

Comment:
`Stream.Read` might [read fewer bytes](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=netframework-4.6.2#System_IO_Stream_Read_System_Byte___System_Int32_System_Int32_) than what is requested.  It would be better to use `Utility.ForceStreamRead` to ensure that the requested number of bytes is actually read.